File: app/src/main/java/com/xdk78/nimebox/ui/MainFragment.java
package com.xdk78.nimebox.ui;

import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.xdk78.nimebox.R;
import com.xdk78.nimebox.adapter.ArticleAdapter;
import com.xdk78.nimebox.api.RepositoryProvider;
import com.xdk78.nimebox.model.Article;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

import static com.xdk78.nimebox.util.Utils.BASE_URL;

public class MainFragment extends Fragment {

    private View view;
    private RecyclerView recyclerView;
    private ArticleAdapter adapter;
    private List<Article> articles;
    private Unbinder unbinder;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public MainFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        view = inflater.inflate(R.layout.fragment_main, container, false);
        unbinder = ButterKnife.bind(this, view);

        recyclerView = view.findViewById(R.id.articles);

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        adapter = new ArticleAdapter(articles = new ArrayList<>(), getContext());
        recyclerView.setAdapter(adapter);
        loadAPI();
        return view;
    }

    public void loadAPI() {
        RepositoryProvider repository = new RepositoryProvider();
        compositeDisposable.add(
                repository.provideRepository(BASE_URL).getArticles()
                        .toList()
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeOn(Schedulers.io())
                        .subscribe(
                                articles -> adapter.addItems(articles),
                                e -> Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show()
                        )
        );
    }

    @Override
    public void onDestroyView() {
        unbinder.unbind();
        compositeDisposable.clear();
        super.onDestroyView();
    }
}

File: app/src/main/java/com/xdk78/nimebox/ui/MainFragment.kt
package com.xdk78.nimebox.ui

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.xdk78.nimebox.R
import com.xdk78.nimebox.adapter.ArticleAdapter
import com.xdk78.nimebox.di.module.api.ApiModule
import com.xdk78.nimebox.di.module.api.ApiRepository
import com.xdk78.nimebox.di.module.api.ApiService
import com.xdk78.nimebox.mvp.base.BaseFragment
import com.xdk78.nimebox.mvp.model.ArticleModel
import com.xdk78.nimebox.mvp.view.ArticleView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import org.jetbrains.anko.support.v4.find
import retrofit2.Retrofit
import java.io.IOException
import javax.inject.Inject

/**
 * Created by xdk78 on 2017-12-02.
 */
class MainFragment : BaseFragment(), ArticleView {

    lateinit var adapter: ArticleAdapter

    private val compositeDisposable: CompositeDisposable = CompositeDisposable()

    @Inject
    lateinit var repository: ApiRepository

    @Inject
    lateinit var apiService: ApiService

    @Inject
    lateinit var retrofit: Retrofit


    private val apiModule: ApiModule = ApiModule()

    private lateinit var recyclerView: RecyclerView

    companion object {
        fun newInstance(): MainFragment {
            return MainFragment()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_main, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
//        // Publish events
//        RxBus.publish(MessageEvent(1, "test"))
        initRetrofit()
        apiService = apiModule.provideApiService(retrofit)
        repository = ApiRepository(apiService)
        loadNews()

    }

    private fun initRetrofit() {
        retrofit = apiModule.provideRetrofit(apiModule.provideRxJavaCallAdapter(), apiModule.provideGsonConverterFactory())
    }

    private fun loadNews() {
        compositeDisposable.add(
                repository
                        .getNews()
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeOn(Schedulers.io())
                        .subscribe(
                                { items -> onItemLoaded(items) },
                                { error -> onError(error) })
        )

    }

    override fun onItemLoaded(items: MutableList<ArticleModel>) {
        recyclerView = find(R.id.articles)
        recyclerView.layoutManager = LinearLayoutManager(context)
        adapter = ArticleAdapter(context, items)
        recyclerView.adapter = adapter
    }

    override fun onError(throwable: Throwable?) {
        if (throwable is IOException) {
            Log.e("api", throwable.message)
        } else {
            Log.i("api", "List is empty")
        }
    }

    override fun onDestroyView() {
        compositeDisposable.clear()
        super.onDestroyView()
    }

}